google/cloud/internal/service_endpoint.cc

```diff
@@ -0,0 +1,57 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/internal/service_endpoint.h"
+#include "google/cloud/internal/absl_str_cat_quiet.h"
+#include "google/cloud/internal/make_status.h"
+#include "google/cloud/universe_domain_options.h"
+#include "absl/strings/strip.h"
+
+namespace google {
+namespace cloud {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace internal {
+
+// TODO(#13122): This function may need to be enhanced to support an env var
+// for UniverseDomain.
+StatusOr<std::string> DetermineServiceEndpoint(
+    absl::optional<std::string> endpoint_env_var,
+    absl::optional<std::string> endpoint_option, Options const& options,
+    std::string default_endpoint) {
+  auto constexpr kGoogleDefaultUniverse = ".googleapis.com.";
```

Comment:
Nit:  I'd probably move this down to immediately before the one place it is (now) used (at which point giving it a name may be overkill)